{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.15.2'"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import regex as re\n",
    "import numpy as np\n",
    "import itertools\n",
    "import operator\n",
    "from collections import Counter\n",
    "from scipy import stats\n",
    "from statistics import mean\n",
    "import gzip\n",
    "import pickle\n",
    "import pysam\n",
    "import mygene\n",
    "pysam.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "bamfile = pysam.AlignmentFile(\"E4_mini_Ig.bam\",\"rb\")\n",
    "no_split = pysam.AlignmentFile(\"No-split.bam\", \"wb\", template=bamfile)\n",
    "split = pysam.AlignmentFile(\"Split.bam\", \"wb\", template=bamfile)\n",
    "n = 0\n",
    "for line in bamfile:\n",
    "    n +=1\n",
    "    if line.has_tag(\"SA\"):\n",
    "        split.write(line)\n",
    "    else:\n",
    "        no_split.write(line)\n",
    "\n",
    "no_split.close()\n",
    "split.close()\n",
    "bamfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "bamfile = pysam.AlignmentFile(\"Split.bam\")\n",
    "tabixfile = pysam.TabixFile(\"coding_gene_region.bed.gz\")\n",
    "out_bam = pysam.AlignmentFile(\"translocations.bam\", \"wb\", template=bamfile)\n",
    "\n",
    "for line in bamfile:\n",
    "\n",
    "    a = line.get_tag(\"SA\").split(\";\")[0]\n",
    "    chrom, start, strand, cigar, mapq, nm = a.split(\",\")\n",
    "    \n",
    "    match = re.findall(r'[0-9]+[A-Za-z]', cigar)\n",
    "    for m in match:\n",
    "        if \"M\" in m:\n",
    "            m = m.replace(\"M\", \"\")\n",
    "            if \"chrUn\" in chrom or \"alt\" in chrom:\n",
    "                pass\n",
    "            else:\n",
    "                for gtf in tabixfile.fetch(chrom, int(start), int(start)+int(m)):\n",
    "                    gtf = gtf.split(\"\\t\")\n",
    "                    if gtf[0] == \"chrM\":\n",
    "                        pass\n",
    "                    else:\n",
    "                        line.tags += [(\"Ta\", gtf[0])]\n",
    "                        line.tags += [(\"Tb\", gtf[1])]\n",
    "                        line.tags += [(\"Tc\", gtf[2])]\n",
    "                        line.tags += [(\"Td\", gtf[3])]\n",
    "\n",
    "                        out_bam.write(line)\n",
    "out_bam.close()\n",
    "bamfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now parse translocations and add information regarding the IG\n",
    "\n",
    "bamfile = pysam.AlignmentFile(\"translocations.bam\",\"r\")\n",
    "tabixfile = pysam.TabixFile(\"Ig_regions.bed.gz\")\n",
    "out_bam = pysam.AlignmentFile(\"translocations_IgAnnot.bam\", \"wb\", template=bamfile)\n",
    "\n",
    "\n",
    "for line in bamfile:\n",
    "\n",
    "\n",
    "    start = line.reference_start\n",
    "    end = line.reference_end\n",
    "    chrom = line.reference_name\n",
    "\n",
    "    for gtf in tabixfile.fetch(chrom, int(start), int(end)):\n",
    "        ig = gtf.split(\"\\t\")[3]\n",
    "        \n",
    "        line.tags += [(\"IG\", ig)]\n",
    "\n",
    "        out_bam.write(line)\n",
    "\n",
    "    \n",
    "tabixfile.close()\n",
    "            \n",
    "\n",
    "    \n",
    "bamfile.close()\n",
    "out_bam.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now need to parse the bam file and generate bed files for each record to make circal plot\n",
    "\n",
    "bamfile = pysam.AlignmentFile(\"translocations_IgAnnot.bam\",\"r\")\n",
    "bed1 = open(\"Ig_features.bed\", \"w\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"w\")\n",
    "\n",
    "\n",
    "for line in bamfile:\n",
    "    \n",
    "    barcode = line.query_name.split(\"_\")[1]\n",
    "\n",
    "    chrom1 = line.reference_name\n",
    "    start1 = line.reference_start\n",
    "    end1 = line.reference_end\n",
    "    ig = line.get_tag(\"IG\")\n",
    "    \n",
    "\n",
    "    chrom2 = line.get_tag(\"Ta\")\n",
    "    start2 = line.get_tag(\"Tb\")\n",
    "    end2 = line.get_tag(\"Tc\")\n",
    "    gene = line.get_tag(\"Td\")\n",
    "\n",
    "\n",
    "    bed1.write(\"%s\\t%s\\t%s\\t%s\\t%s\\n\" % (chrom1, start1, end1, ig, barcode))\n",
    "    bed2.write(\"%s\\t%s\\t%s\\t%s\\t%s\\n\" % (chrom2, start2, end2, gene, barcode))\n",
    "        \n",
    "bed1.close()\n",
    "bed2.close()\n",
    "bamfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# %%timeit\n",
    "import collections\n",
    "\n",
    "bed1 = open(\"Ig_features.bed\", \"r\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"r\")\n",
    "\n",
    "out_table = open(\"annotated_trans_ens.txt\", \"w\")\n",
    "\n",
    "trans_list = []\n",
    "\n",
    "#gc.disable()\n",
    "n = 0\n",
    "for bed1, bed2 in zip(bed1, bed2):\n",
    "\n",
    "    barcode = bed1.split(\"\\t\")[4]\n",
    "    ig = bed1.split(\"\\t\")[3]\n",
    "    transloc = bed2.split(\"\\t\")[3]\n",
    "    trans_gene = ig + \"_\" + transloc + \"_\"+ barcode.strip()\n",
    "    trans_list.append(str(trans_gene))\n",
    "#gc.enable()\n",
    "out_table.write(\"Ig\\tgene\\ttrans\\tbarcode\\tcount\\n\")\n",
    "print(trans_list)\n",
    "trans_counter =collections.Counter(trans_list)\n",
    "print(trans_counter)\n",
    "for k in trans_counter:\n",
    "    ig, gene, barcode = k.split(\"_\")\n",
    "    gene2 = ig + \"_\" + gene\n",
    "    counter = trans_counter[k]\n",
    "\n",
    "    out_table.write(\"%s\\t%s\\t%s\\t%s\\t%s\\n\"%(ig, gene, gene2, barcode, str(counter)))\n",
    "    \n",
    "out_table.close()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = open(\"annotated_trans_ens.txt\", \"r\")\n",
    "barcodes = open(\"jjn3_barcodes.csv\", \"r\")\n",
    "outf = open(\"JJN3_counts.txt\",\"w\")\n",
    "barcodes_list = []\n",
    "\n",
    "for i in barcodes:\n",
    "    barcodes_list.append(i.strip())\n",
    "    \n",
    "n = 0\n",
    "for i in infile:\n",
    "\n",
    "    barcode = i.strip().split(\"\\t\")[3]\n",
    "    if barcode in barcodes_list:\n",
    "        outf.write(\"%s\"%(i))\n",
    "\n",
    "            \n",
    "outf.close()\n",
    "infile.close()\n",
    "barcodes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = open(\"annotated_trans_ens.txt\", \"r\")\n",
    "barcodes = open(\"df15_barcodes.csv\", \"r\")\n",
    "outf = open(\"DF15_counts.txt\",\"w\")\n",
    "barcodes_list = []\n",
    "\n",
    "for i in barcodes:\n",
    "    barcodes_list.append(i.strip())\n",
    "    \n",
    "n = 0\n",
    "for i in infile:\n",
    "\n",
    "    barcode = i.strip().split(\"\\t\")[3]\n",
    "    if barcode in barcodes_list:\n",
    "        outf.write(\"%s\"%(i))\n",
    "            \n",
    "            \n",
    "outf.close()\n",
    "infile.close()\n",
    "barcodes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = open(\"annotated_trans_ens.txt\", \"r\")\n",
    "barcodes = open(\"h929_barcodes.csv\", \"r\")\n",
    "outf = open(\"H929_counts.txt\",\"w\")\n",
    "barcodes_list = []\n",
    "\n",
    "for i in barcodes:\n",
    "    barcodes_list.append(i.strip())\n",
    "    \n",
    "n = 0\n",
    "for i in infile:\n",
    "\n",
    "    barcode = i.strip().split(\"\\t\")[3]\n",
    "    if barcode in barcodes_list:\n",
    "        outf.write(\"%s\"%(i))\n",
    "            \n",
    "            \n",
    "outf.close()\n",
    "infile.close()\n",
    "barcodes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed1 = open(\"Ig_features.bed\", \"r\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"r\")\n",
    "barcode = open(\"jjn3_barcodes.csv\", \"r\")\n",
    "outf = open(\"Ig_features_JJN3.bed\",\"w\")\n",
    "outf2 = open(\"Translocation_features_JJN3.bed\",\"w\")\n",
    "\n",
    "barcode_list = []\n",
    "\n",
    "for i in barcode:\n",
    "    barcode_list.append(i.strip())\n",
    "\n",
    "for bed1, bed2 in zip(bed1, bed2):\n",
    "\n",
    "    barcode = bed1.split(\"\\t\")[4]\n",
    "    ig = bed1.split(\"\\t\")[3]\n",
    "    barcode = barcode.strip()\n",
    "    \n",
    "    if barcode in barcode_list:\n",
    "        outf.write(bed1)\n",
    "        outf2.write(bed2)\n",
    "        \n",
    "outf.close()\n",
    "outf2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed1 = open(\"Ig_features.bed\", \"r\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"r\")\n",
    "barcode = open(\"df15_barcodes.csv\", \"r\")\n",
    "outf = open(\"Ig_features_DF15.bed\",\"w\")\n",
    "outf2 = open(\"Translocation_features_DF15.bed\",\"w\")\n",
    "\n",
    "barcode_list = []\n",
    "\n",
    "for i in barcode:\n",
    "    barcode_list.append(i.strip())\n",
    "\n",
    "for bed1, bed2 in zip(bed1, bed2):\n",
    "\n",
    "    barcode = bed1.split(\"\\t\")[4]\n",
    "    ig = bed1.split(\"\\t\")[3]\n",
    "    barcode = barcode.strip()\n",
    "    \n",
    "    if barcode in barcode_list:\n",
    "        outf.write(bed1)\n",
    "        outf2.write(bed2)\n",
    "        \n",
    "outf.close()\n",
    "outf2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed1 = open(\"Ig_features.bed\", \"r\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"r\")\n",
    "barcode = open(\"h929_barcodes.csv\", \"r\")\n",
    "outf = open(\"Ig_features_H929.bed\",\"w\")\n",
    "outf2 = open(\"Translocation_features_H929.bed\",\"w\")\n",
    "\n",
    "barcode_list = []\n",
    "\n",
    "for i in barcode:\n",
    "    barcode_list.append(i.strip())\n",
    "\n",
    "for bed1, bed2 in zip(bed1, bed2):\n",
    "\n",
    "    barcode = bed1.split(\"\\t\")[4]\n",
    "    ig = bed1.split(\"\\t\")[3]\n",
    "    barcode = barcode.strip()\n",
    "    \n",
    "    if barcode in barcode_list:\n",
    "        outf.write(bed1)\n",
    "        outf2.write(bed2)\n",
    "        \n",
    "outf.close()\n",
    "outf2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed1 = open(\"Ig_features.bed\", \"r\")\n",
    "bed2 = open(\"Translocation_features.bed\", \"r\")\n",
    "barcode = open(\"total_barcodes.csv\", \"r\")\n",
    "outf = open(\"Ig_features_total.bed\",\"w\")\n",
    "outf2 = open(\"Translocation_features_total.bed\",\"w\")\n",
    "\n",
    "barcode_list = []\n",
    "\n",
    "for i in barcode:\n",
    "    barcode_list.append(i.strip())\n",
    "\n",
    "for bed1, bed2 in zip(bed1, bed2):\n",
    "\n",
    "    barcode = bed1.split(\"\\t\")[4]\n",
    "    ig = bed1.split(\"\\t\")[3]\n",
    "    barcode = barcode.strip()\n",
    "    \n",
    "    if barcode in barcode_list:\n",
    "        outf.write(bed1)\n",
    "        outf2.write(bed2)\n",
    "        \n",
    "outf.close()\n",
    "outf2.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
